### PR TITLE
Add padding for Navigation Bar to account for safe area

### LINKED ISSUE #101605: NavigationBar doesn't account for safe area
Details in b/228601418. Seems like NavigationBar doesn't internally account for horizontal safe area, which becomes an issue with notches etc when in landscape mode. 

A release blocker for customer, though release is still a few months away.

### PR DESCRIPTION
**Pull request by QuncCccccc**
This PR adds paddings for Navigation Bar so that when a Android/iOS device is in landscape mode, the Navigation Bar will respect SafeAreas by default. 

Fixes #101605.
Screen Shot 2022-04-22 at 5 53 48 PM
Screen Shot 2022-04-22 at 5 55 24 PM
Screen Shot 2022-04-22 at 5 56 15 PM

## Pre-launch Checklist
- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: packages/flutter/lib/src/material/navigation_bar.dart [hunk 1]
```diff
@@ -148,14 +148,17 @@ class NavigationBar extends StatelessWidget {
...
         ?? navigationBarTheme.backgroundColor
         ?? defaults.backgroundColor!,
       elevation: elevation ?? navigationBarTheme.elevation ?? defaults.elevation!,
       child: Padding(
-        padding: EdgeInsets.only(bottom: additionalBottomPadding),
+        padding: EdgeInsets.fromLTRB(additionalLeftPadding, 0,
+          additionalRightPadding, additionalBottomPadding),
         child: MediaQuery.removePadding(
           context: context,
           removeBottom: true,
```
Comment (AlexV525) [inline packages/flutter/lib/src/material/navigation_bar.dart:164]: Should we set `removeLeft` and `removeRight` here?

Comment (goderbauer) [reply packages/flutter/lib/src/material/navigation_bar.dart:164]: Looks like that padding is getting consumed above, so it should be removed for ancestors as suggested.

File: packages/flutter/test/material/navigation_bar_test.dart [hunk 2]
```diff
@@ -137,6 +138,87 @@ void main() {
...
+
+    // The position of center item of navigation bar should indicate whether
+    // the safe area is sufficiently respected, when safe areas are on both sides.
+    // e.g. iOS device with both sides of round corner.
+    final Finder paddedCenterItem = find.byKey(const Key('Center'));
+    final Offset paddedCenter = tester.getCenter(paddedCenterItem);
+    expect(paddedCenter.dx, closeTo(defaultWidth / 2, precisionErrorTolerance));
+  });
+
+
```
Comment (goderbauer) [inline packages/flutter/test/material/navigation_bar_test.dart:221]: nit: remove extra whitespace here.

File: packages/flutter/lib/src/material/navigation_bar.dart (hunk 1 above)
Comment (QuncCccccc) [reply packages/flutter/lib/src/material/navigation_bar.dart]: Thanks for the comments! Looks like SafeArea did same thing just like Padding() and MediaQuery.removePadding(), so I removed both and use SafeArea instead. Let me know if you have any suggestions.
https://github.com/flutter/flutter/blob/5464c5bac7/packages/flutter/lib/src/widgets/safe_area.dart#:~:text=return%20Padding(,child%3A%20MediaQuery.removePadding(

File: packages/flutter/test/material/navigation_bar_test.dart [hunk 3]
```diff
@@ -137,6 +138,86 @@ void main() {
     expect(tester.getSize(find.byType(NavigationBar)).height, expectedHeight);
   });
 
+  testWidgets('NavigationBar respects the notch/system navigation bar in landscape mode', (WidgetTester tester) async {
+    const double safeAreaPadding = 40.0;
+    Widget _navigationBar() {
```
Comment (goderbauer) [inline packages/flutter/test/material/navigation_bar_test.dart:143]: nit: remove underscore for local functions.

File: packages/flutter/test/material/navigation_bar_test.dart [hunk 4]
```diff
@@ -137,6 +138,86 @@ void main() {
...
+      ),
+    );
+
+    // The position of center item of navigation bar should indicate whether
+    // the safe area is sufficiently respected, when safe area is on the right side.
+    // e.g. Android device with system navigation bar in landscape mode.
+    final Finder rightPaddedCenterItem = find.byKey(const Key('Center'));
+    final Offset rightPaddedCenter = tester.getCenter(rightPaddedCenterItem);
+    expect(rightPaddedCenter.dx, closeTo((defaultWidth - safeAreaPadding) / 2,
+        precisionErrorTolerance));
```
Comment (goderbauer) [inline packages/flutter/test/material/navigation_bar_test.dart:201]: format nit to make this easier to read:

```suggestion
    expect(
      rightPaddedCenter.dx, 
      closeTo((defaultWidth - safeAreaPadding) / 2, precisionErrorTolerance),
    );
```

Basically, the line with the opening and closing parenthesis should always have the same indentation.

File: packages/flutter/test/material/navigation_bar_test.dart [hunk 5]
```diff
@@ -137,6 +138,86 @@ void main() {
...
+
+    await tester.pumpWidget(
+      _buildWidget(
+        MediaQuery(
+          data: const MediaQueryData(padding: EdgeInsets.fromLTRB(
+              safeAreaPadding, 0, safeAreaPadding, safeAreaPadding)),
```
Comment (goderbauer) [inline packages/flutter/test/material/navigation_bar_test.dart:207]: format nit: properties of a widget should be on their own line, e.g.

```suggestion
          data: const MediaQueryData(
            padding: EdgeInsets.fromLTRB(
              safeAreaPadding, 0, safeAreaPadding, safeAreaPadding,
            ),
          ),
```

**Review by goderbauer — APPROVED** [review]
LGTM with some formatting nits

File: packages/flutter/test/material/navigation_bar_test.dart (hunk 5 above)
Comment (AlexV525) [inline packages/flutter/test/material/navigation_bar_test.dart:207]: I would prefer:

```suggestion
          data: const MediaQueryData(
            padding: EdgeInsets.fromLTRB(
              safeAreaPadding,
              0,
              safeAreaPadding,
              safeAreaPadding,
            ),
          ),
```

File: packages/flutter/lib/src/material/navigation_bar.dart [hunk 6]
```diff
@@ -147,41 +147,35 @@ class NavigationBar extends StatelessWidget {
...
     return Material(
       color: backgroundColor
         ?? navigationBarTheme.backgroundColor
         ?? defaults.backgroundColor!,
       elevation: elevation ?? navigationBarTheme.elevation ?? defaults.elevation!,
-      child: Padding(
-        padding: EdgeInsets.only(bottom: additionalBottomPadding),
-        child: MediaQuery.removePadding(
-          context: context,
-          removeBottom: true,
```
Comment (darrenaustin) [inline packages/flutter/lib/src/material/navigation_bar.dart:161]: Do we know what this padding and MediaQuery were doing here? It looks like they were doing something specific with bottom padding that has been removed in this PR.

File: packages/flutter/test/material/navigation_bar_test.dart [hunk 7]
```diff
@@ -137,6 +138,107 @@ void main() {
...
+          NavigationDestination(
+            icon: Icon(Icons.access_alarm),
+            label: 'Alarm',
+          ),
+        ],
+        onDestinationSelected: (int i) {},
+      );
+    }
+
+    await tester.pumpWidget(_buildWidget(navigationBar(),));
```
Comment (darrenaustin) [inline packages/flutter/test/material/navigation_bar_test.dart:164]: Nit: don't need the extra ','.

File: packages/flutter/test/material/navigation_bar_test.dart [hunk 8]
```diff
@@ -137,6 +138,107 @@ void main() {
...
+    // the safe area is sufficiently respected, when safe area is on the left side.
+    // e.g. Android device with system navigation bar in landscape mode.
+    final Finder leftPaddedCenterItem = find.byKey(const Key('Center'));
+    final Offset leftPaddedCenter = tester.getCenter(leftPaddedCenterItem);
+    expect(
+        leftPaddedCenter.dx,
+        closeTo(
+            (defaultWidth + safeAreaPadding) / 2.0,
+            precisionErrorTolerance
+        ),
```
Comment (darrenaustin) [inline packages/flutter/test/material/navigation_bar_test.dart:191]: nit: we usually indent only two spaces for continued lines like this. This applies to all the code below this as well. 

In this case it looks like the `closeTo` call could also be on one line?

**Review by darrenaustin — COMMENTED** [review]
Looks good. Just a question and a couple of nits below.

File: packages/flutter/lib/src/material/navigation_bar.dart (hunk 6 above)
Comment (QuncCccccc) [reply packages/flutter/lib/src/material/navigation_bar.dart:161]: Based on my understanding, the Padding(...MediaQuery.removePadding(...)) part is just giving a SafeArea for device bottom. The code does same thing just like what SafeArea api does, so when I want to add additionalLeftPadding and additionalRightPadding, I replaced all of them with SafeArea. Please let me know if it makes sense to you:)

**Review by darrenaustin — APPROVED** [review]
LGTM.

**Comment by Moncader** [comment]
@QuncCccccc @darrenaustin 
This PR has made a degradation in Flutter 3.3 where if you have an app that doesn't use a widget consuming the MediaQuery's top padding (like AppBar does), the top padding of the safe area (the notification shade area) gets added to the top of the NavigationBar making the bar very tall

**Pull request merged**